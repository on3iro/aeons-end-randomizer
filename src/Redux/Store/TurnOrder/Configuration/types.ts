import { ActionsUnion } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { actions } from './actions'

export type State = Readonly<{
  Mode: types.Mode
  SelectedPlayerCount: types.ITurnOrderPlayerCount
  SelectedSetup: types.ITurnOrderSetup
}>

export enum ActionTypes {
  SET_MODE = 'TurnOrder/Configuration/SET_MODE',
  SELECT_PLAYER_COUNT = 'TurnOrder/Configuration/SELECT_PLAYER_COUNT',
  SELECT_SETUP = 'TurnOrder/Configuration/SELECT_SETUP',
  SET_TO_DB = 'TurnOrder/Configuration/SET_TO_DB',
  SET_TO_DB_SUCCESS = 'TurnOrder/Configuration/SET_TO_DB_SUCCESS',
  SET_TO_DB_FAILURE = 'TurnOrder/Configuration/SET_TO_DB_FAILURE',
  FETCH_FROM_DB = 'TurnOrder/Configuration/FETCH_FROM_DB',
  FETCH_FROM_DB_SUCCESS = 'TurnOrder/Configuration/FETCH_FROM_DB_SUCCESS',
  FETCH_FROM_DB_FAILURE = 'TurnOrder/Configuration/FETCH_FROM_DB_FAILURE',
}

export type Action = ActionsUnion<typeof actions>

export type ModeStateSlice = {
  TurnOrder: {
    Configuration: {
      Mode: types.Mode
    }
  }
}

export type SelectedPlayerCountStateSlice = {
  TurnOrder: {
    Configuration: {
      SelectedPlayerCount: types.ITurnOrderPlayerCount
    }
  }
}

export type SelectedSetupStateSlice = {
  TurnOrder: {
    Configuration: {
      SelectedSetup: types.ITurnOrderSetup
    }
  }
}
