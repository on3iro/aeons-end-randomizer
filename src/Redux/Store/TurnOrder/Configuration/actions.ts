import { createAction } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { ActionTypes } from './types'

export const actions = {
  noOp: () => createAction('@@REDUX_LOOP/ENFORCE_DEFAULT_HANDLING'),
  setMode: (mode: types.Mode) => createAction(ActionTypes.SET_MODE, mode),
  selectPlayerCount: (playerCountId: types.PlayerCountId) =>
    createAction(ActionTypes.SELECT_PLAYER_COUNT, playerCountId),
  selectSetup: (setupId: types.TurnorderSetupVariationId) =>
    createAction(ActionTypes.SELECT_SETUP, setupId),
  setToDB: () => createAction(ActionTypes.SET_TO_DB),
  setToDBSuccessful: () => createAction(ActionTypes.SET_TO_DB_SUCCESS),
  setToDBFailed: () => createAction(ActionTypes.SET_TO_DB_FAILURE),
  fetchFromDB: () => createAction(ActionTypes.FETCH_FROM_DB),
  fetchFromDBSuccessful: (state: unknown) =>
    createAction(ActionTypes.FETCH_FROM_DB_SUCCESS, state),
  fetchFromDBFailed: () => createAction(ActionTypes.FETCH_FROM_DB_FAILURE),
}
