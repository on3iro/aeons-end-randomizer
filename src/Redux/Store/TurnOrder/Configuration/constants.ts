import AERData from 'aer-data/src/index'

const INITIAL_MODE = 'Default'
const INITIAL_PLAYER_SETUP = AERData.turnordersetups['onePlayer']
const INITIAL_VARIATION = INITIAL_PLAYER_SETUP.variations['default']

const TURNORDER_CONFIG_DB_KEY = '2-4-turnOrderConfiguration'

export {
  INITIAL_MODE,
  INITIAL_PLAYER_SETUP,
  INITIAL_VARIATION,
  TURNORDER_CONFIG_DB_KEY,
}
