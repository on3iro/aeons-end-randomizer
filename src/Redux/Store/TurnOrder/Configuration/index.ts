import { createAction, ActionsUnion } from '@martin_hotell/rex-tils'
import { LoopReducer, loop, Cmd } from 'redux-loop'
import { createSelector } from 'reselect'
import { get as getFromDb, set as setToDb } from 'idb-keyval'

import config from '../../../../config'
import { ITurnOrderSetup, Mode } from '../../../../types'
import { RootState } from '../../'
import {
  actions as activeGameActions,
  Action as ActiveGameAction,
} from '../ActiveGame'

const TURNORDER_CONFIG_DB_KEY = 'turnOrderConfig'

/////////////
// HELPERS //
/////////////

const adjustSetup = (mode: Mode, setup: ITurnOrderSetup): ITurnOrderSetup => {
  switch (mode) {
    case 'Maelstrom': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          return card.id === 'nemesis-1' || card.id === 'blitz'
            ? config.TURNORDERCARDS['maelstrom']
            : card
        }),
      }
    }

    case 'Blitz': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          return card.id === 'nemesis-1' || card.id === 'maelstrom'
            ? config.TURNORDERCARDS['blitz']
            : card
        }),
      }
    }

    case 'Default':
    default: {
      return setup
    }
  }
}

///////////
// STATE //
///////////

export type State = Readonly<{
  Mode: Mode
  SelectedSetup: ITurnOrderSetup
}>
export const initialState: State = {
  Mode: 'Default',
  SelectedSetup: config.TURNORDERSETUPS['onePlayerThreeToc'],
}

/////////////
// ACTIONS //
/////////////

export enum ActionTypes {
  SET_MODE = 'TurnOrder/Configuration/SET_MODE',
  SELECT_SETUP = 'TurnOrder/Configuration/SELECT_SETUP',
  SET_TO_DB = 'TurnOrder/Configuration/SET_TO_DB',
  SET_TO_DB_SUCCESS = 'TurnOrder/Configuration/SET_TO_DB_SUCCESS',
  SET_TO_DB_FAILURE = 'TurnOrder/Configuration/SET_TO_DB_FAILURE',
  FETCH_FROM_DB = 'TurnOrder/Configuration/FETCH_FROM_DB',
  FETCH_FROM_DB_SUCCESS = 'TurnOrder/Configuration/FETCH_FROM_DB_SUCCESS',
  FETCH_FROM_DB_FAILURE = 'TurnOrder/Configuration/FETCH_FROM_DB_FAILURE',
}

export const actions = {
  noOp: () => createAction('NOOP'),
  setMode: (mode: Mode) => createAction(ActionTypes.SET_MODE, mode),
  selectSetup: (setupId: string) =>
    createAction(ActionTypes.SELECT_SETUP, setupId),
  setToDB: () => createAction(ActionTypes.SET_TO_DB),
  setToDBSuccessful: () => createAction(ActionTypes.SET_TO_DB_SUCCESS),
  setToDBFailed: () => createAction(ActionTypes.SET_TO_DB_FAILURE),
  fetchFromDB: () => createAction(ActionTypes.FETCH_FROM_DB),
  fetchFromDBSuccessful: (state: State) =>
    createAction(ActionTypes.FETCH_FROM_DB_SUCCESS, state),
  fetchFromDBFailed: () => createAction(ActionTypes.FETCH_FROM_DB_FAILURE),
}

export type Action = ActionsUnion<typeof actions>

/////////////
// REDUCER //
/////////////

export const Reducer: LoopReducer<State, Action | ActiveGameAction> = (
  state: State = initialState,
  action: Action
) => {
  switch (action.type) {
    case ActionTypes.SET_MODE: {
      const newState = { ...state, Mode: action.payload }
      return loop(
        newState,
        Cmd.run<Action>(setToDb, {
          args: [TURNORDER_CONFIG_DB_KEY, newState],
          successActionCreator: actions.setToDBSuccessful,
          failActionCreator: actions.setToDBFailed,
        })
      )
    }

    case ActionTypes.SELECT_SETUP: {
      const newState = {
        ...state,
        SelectedSetup: config.TURNORDERSETUPS[action.payload],
      }
      return loop(
        newState,
        Cmd.run<Action>(setToDb, {
          args: [TURNORDER_CONFIG_DB_KEY, newState],
          successActionCreator: actions.setToDBSuccessful,
          failActionCreator: actions.setToDBFailed,
        })
      )
    }

    case ActionTypes.FETCH_FROM_DB: {
      return loop(
        state,
        Cmd.run<Action>(getFromDb, {
          args: [TURNORDER_CONFIG_DB_KEY],
          successActionCreator: actions.fetchFromDBSuccessful,
          failActionCreator: actions.fetchFromDBFailed,
        })
      )
    }

    case ActionTypes.FETCH_FROM_DB_SUCCESS: {
      // If the fetched state somehow is undefined just take the current state instead
      const newState = action.payload || state
      return loop(
        newState,
        // If we could load the configuration also load the actual
        // turnorder game state
        Cmd.action<ActiveGameAction>(activeGameActions.fetchFromDB())
      )
    }

    default: {
      return state
    }
  }
}

///////////////
// SELECTORS //
///////////////

const getMode = (state: RootState) => state.TurnOrder.Configuration.Mode
const getSelectedSetup = (state: RootState) =>
  state.TurnOrder.Configuration.SelectedSetup
const getConfiguration = createSelector(
  [getMode, getSelectedSetup],
  (mode, selectedSetup) => adjustSetup(mode, selectedSetup)
)
const getAvailableCards = createSelector(
  [getConfiguration],
  config => config.turnOrderCards
)

export const selectors = {
  getMode,
  getSelectedSetup,
  getConfiguration,
  getAvailableCards,
}
