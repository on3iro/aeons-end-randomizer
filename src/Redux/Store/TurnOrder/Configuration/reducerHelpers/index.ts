export * from './setMode'
export * from './selectPlayerCount'
export * from './selectSetup'
export * from './fetchFromDb'
