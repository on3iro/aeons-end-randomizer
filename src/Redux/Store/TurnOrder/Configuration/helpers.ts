import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'

import AERData from 'aer-data/src/index'
import * as types from 'aer-types/types'

import { State } from './types'
import { TURNORDER_CONFIG_DB_KEY } from './constants'
import { actions } from './actions'

const newStateWithDBWrite = (newState: State) => {
  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [TURNORDER_CONFIG_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}

const adjustSetup = (
  mode: types.Mode,
  setup: types.ITurnOrderSetup
): types.ITurnOrderSetup => {
  switch (mode) {
    case 'Maelstrom': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          return card.id === 'nemesis-1' || card.id === 'blitz'
            ? AERData.turnordercards['maelstrom']
            : card
        }),
      }
    }

    case 'Blitz': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          return card.id === 'nemesis-1' || card.id === 'maelstrom'
            ? AERData.turnordercards['blitz']
            : card
        }),
      }
    }

    case 'Blitz + Maelstrom': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          if (card.id === 'nemesis-1') {
            return AERData.turnordercards['blitz']
          }

          if (card.id === 'nemesis-2') {
            return AERData.turnordercards['maelstrom']
          }

          return card
        }),
      }
    }

    case 'Thief Of Dreams': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          return card.id === 'nemesis-1'
            ? AERData.turnordercards['thiefOfDreams']
            : card
        }),
      }
    }

    case 'Paradox of Myth and Bone': {
      return {
        id: setup.id,
        name: setup.name,
        turnOrderCards: setup.turnOrderCards.map(card => {
          if (card.id === 'nemesis-1') {
            return AERData.turnordercards['myth']
          }

          if (card.id === 'nemesis-2') {
            return AERData.turnordercards['bone']
          }

          return card
        }),
      }
    }

    case 'Default':
    default: {
      return setup
    }
  }
}

export { newStateWithDBWrite, adjustSetup }
