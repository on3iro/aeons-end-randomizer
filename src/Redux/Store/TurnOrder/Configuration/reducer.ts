import { Action, ActionTypes, State } from './types'
import {
  INITIAL_MODE,
  INITIAL_PLAYER_SETUP,
  INITIAL_VARIATION,
} from './constants'

import * as reducerHelpers from './reducerHelpers'

export const initialState: State = {
  Mode: INITIAL_MODE,
  SelectedPlayerCount: INITIAL_PLAYER_SETUP,
  SelectedSetup: INITIAL_VARIATION,
}

export const Reducer = (state: State = initialState, action: Action) => {
  switch (action.type) {
    case ActionTypes.SET_MODE: {
      return reducerHelpers.setMode(state, action)
    }

    case ActionTypes.SELECT_PLAYER_COUNT: {
      return reducerHelpers.selectPlayerCount(state, action)
    }

    case ActionTypes.SELECT_SETUP: {
      return reducerHelpers.selectSetup(state, action)
    }

    case ActionTypes.FETCH_FROM_DB: {
      return reducerHelpers.fetchFromDb(state)
    }

    case ActionTypes.FETCH_FROM_DB_SUCCESS: {
      return reducerHelpers.fetchFromDbSuccess(state, action)
    }

    case ActionTypes.FETCH_FROM_DB_FAILURE: {
      return state
    }

    default: {
      return state
    }
  }
}
