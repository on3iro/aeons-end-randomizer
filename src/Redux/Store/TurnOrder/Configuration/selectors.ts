import { createSelector } from 'reselect'

import {
  ModeStateSlice,
  SelectedPlayerCountStateSlice,
  SelectedSetupStateSlice,
} from './types'

import { adjustSetup } from './helpers'

const getMode = (state: ModeStateSlice) => state.TurnOrder.Configuration.Mode

const getSelectedPlayerCount = (state: SelectedPlayerCountStateSlice) =>
  state.TurnOrder.Configuration.SelectedPlayerCount

const getSelectedSetup = (state: SelectedSetupStateSlice) =>
  state.TurnOrder.Configuration.SelectedSetup

const getConfiguration = createSelector(
  [getMode, getSelectedSetup],
  (mode, selectedSetup) => adjustSetup(mode, selectedSetup)
)

const getAvailableCards = createSelector(
  [getConfiguration],
  config => config.turnOrderCards
)

export const selectors = {
  getMode,
  getSelectedPlayerCount,
  getSelectedSetup,
  getConfiguration,
  getAvailableCards,
}
