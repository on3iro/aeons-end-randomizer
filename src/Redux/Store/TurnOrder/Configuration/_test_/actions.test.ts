import configureStore from 'redux-mock-store'

import config from 'config'

import { actions } from 'Redux/Store/TurnOrder/Configuration/actions'

const mockStore = configureStore()
const store = mockStore()

describe('Turn Order | Configuration | actions', () => {
  beforeEach(() => {
    store.clearActions()
  })

  it('should dispatch noop action', () => {
    store.dispatch(actions.noOp())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch setMode action with payload', () => {
    store.dispatch(actions.setMode('Default'))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch selectPlayerCount action with payload', () => {
    store.dispatch(actions.selectPlayerCount('twoPlayers'))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch selectSetup action with payload', () => {
    store.dispatch(actions.selectSetup('default'))
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch fetchFromDB action', () => {
    store.dispatch(actions.fetchFromDB())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch fetchFromDBSuccessful action', () => {
    store.dispatch(
      actions.fetchFromDBSuccessful({
        Mode: 'Default',
        SelectedPlayerCount: config.TURNORDERSETUPS['onePlayer'],
        SelectedSetup:
          config.TURNORDERSETUPS['onePlayer'].variations['default'],
      })
    )
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch fetchFromDBFailed action', () => {
    store.dispatch(actions.fetchFromDBFailed())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch setToDB action', () => {
    store.dispatch(actions.setToDB())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch setToDBSuccessful action', () => {
    store.dispatch(actions.setToDBSuccessful())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should dispatch setToDBFailed action', () => {
    store.dispatch(actions.setToDBFailed())
    expect(store.getActions()).toMatchSnapshot()
  })
})
