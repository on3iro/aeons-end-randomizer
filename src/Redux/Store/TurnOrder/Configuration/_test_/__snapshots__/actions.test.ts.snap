// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Turn Order | Configuration | actions should dispatch fetchFromDB action 1`] = `
Array [
  Object {
    "type": "TurnOrder/Configuration/FETCH_FROM_DB",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch fetchFromDBFailed action 1`] = `
Array [
  Object {
    "type": "TurnOrder/Configuration/FETCH_FROM_DB_FAILURE",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch fetchFromDBSuccessful action 1`] = `
Array [
  Object {
    "payload": Object {
      "Mode": "Default",
      "SelectedPlayerCount": Object {
        "id": "onePlayer",
        "name": "1",
        "variations": Object {
          "34alternating": Object {
            "id": "34alternating",
            "name": "3/4 Cards alternating",
            "turnOrderCards": Array [
              Object {
                "id": "player1-1",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-2",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-3",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "alternate": true,
                "id": "player-1-alternating",
                "name": "Player 1 (alternate)",
                "type": "player1",
              },
              Object {
                "id": "nemesis-1",
                "name": "Nemesis",
                "type": "nemesis",
              },
              Object {
                "id": "nemesis-2",
                "name": "Nemesis",
                "type": "nemesis",
              },
            ],
          },
          "4toc": Object {
            "id": "4toc",
            "name": "4 ToC",
            "turnOrderCards": Array [
              Object {
                "id": "player1-1",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-2",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-3",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-4",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "nemesis-1",
                "name": "Nemesis",
                "type": "nemesis",
              },
              Object {
                "id": "nemesis-2",
                "name": "Nemesis",
                "type": "nemesis",
              },
            ],
          },
          "default": Object {
            "id": "default",
            "name": "Default (3 ToC)",
            "turnOrderCards": Array [
              Object {
                "id": "player1-1",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-2",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "player1-3",
                "name": "Player 1",
                "type": "player1",
              },
              Object {
                "id": "nemesis-1",
                "name": "Nemesis",
                "type": "nemesis",
              },
              Object {
                "id": "nemesis-2",
                "name": "Nemesis",
                "type": "nemesis",
              },
            ],
          },
        },
      },
      "SelectedSetup": Object {
        "id": "default",
        "name": "Default (3 ToC)",
        "turnOrderCards": Array [
          Object {
            "id": "player1-1",
            "name": "Player 1",
            "type": "player1",
          },
          Object {
            "id": "player1-2",
            "name": "Player 1",
            "type": "player1",
          },
          Object {
            "id": "player1-3",
            "name": "Player 1",
            "type": "player1",
          },
          Object {
            "id": "nemesis-1",
            "name": "Nemesis",
            "type": "nemesis",
          },
          Object {
            "id": "nemesis-2",
            "name": "Nemesis",
            "type": "nemesis",
          },
        ],
      },
    },
    "type": "TurnOrder/Configuration/FETCH_FROM_DB_SUCCESS",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch noop action 1`] = `
Array [
  Object {
    "type": "NOOP",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch selectPlayerCount action with payload 1`] = `
Array [
  Object {
    "payload": "twoPlayers",
    "type": "TurnOrder/Configuration/SELECT_PLAYER_COUNT",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch selectSetup action with payload 1`] = `
Array [
  Object {
    "payload": "default",
    "type": "TurnOrder/Configuration/SELECT_SETUP",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch setMode action with payload 1`] = `
Array [
  Object {
    "payload": "Default",
    "type": "TurnOrder/Configuration/SET_MODE",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch setToDB action 1`] = `
Array [
  Object {
    "type": "TurnOrder/Configuration/SET_TO_DB",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch setToDBFailed action 1`] = `
Array [
  Object {
    "type": "TurnOrder/Configuration/SET_TO_DB_FAILURE",
  },
]
`;

exports[`Turn Order | Configuration | actions should dispatch setToDBSuccessful action 1`] = `
Array [
  Object {
    "type": "TurnOrder/Configuration/SET_TO_DB_SUCCESS",
  },
]
`;
